File: 백준 15651 N과 M(3) -백트랙킹-/복습/Main.java
import java.util.Scanner;

public class Main {
    static boolean[] isVisited;
    static int[] num;
    static int n;
    static int m;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        isVisited = new boolean[m + 1];
        num = new int[m + 1];
        dfs(1);
        System.out.println(sb);
    }

    static void dfs(int current) {
        if (current > m) {
            for (int i = 1; i <= m; i++) {
                sb.append(num[i]).append(" ");
            }
            sb.append("\n");
        } else {
            for (int i = 1; i <= n; i++) {
                num[current] = i;
                dfs(current + 1);
            }
        }
    }
}

File: 백준 15651 N과 M(3) -백트랙킹-/복습/Main.kt
import java.util.*

internal var isVisited: BooleanArray? = null
internal var num: IntArray? = null
internal var n: Int = 0
internal var m: Int = 0
internal var sb = StringBuilder()

fun main(args: Array<String>) {
    val sc = Scanner(System.`in`)
    n = sc.nextInt()
    m = sc.nextInt()
    isVisited = BooleanArray(m + 1)
    num = IntArray(m + 1)
    dfs(1)
    println(sb)
}

internal fun dfs(current: Int) {
    if (current > m) {
        for (i in 1..m) {
            sb.append(num?.get(i)).append(" ")
        }
        sb.append("\n")
    } else {
        for (i in 1..n) {
            num?.set(current, i)
            dfs(current + 1)
        }
    }
}